chutney/server/src/main/java/com/chutneytesting/execution/domain/purge/PurgeServiceImpl.java

```diff
@@ -145,28 +146,46 @@ private PurgeExecutionService<Campaign, Long, CampaignExecution> buildCampaignSe
             cer -> cer.executionEnvironment,
             campaignExecutionRepository::deleteExecutions
         ) {
+            // Not thread safe
             private Map<Boolean, List<CampaignExecution>> campaignExecutionsByPartialExecution;
+            private List<CampaignExecution> emptyCampaignExecutions;
 
             /**
-             * Transform executions to filter only those that are not manual replays.
+             * Transform executions to filter only those that are not manual replays and not empty.
              */
             @Override
             public List<CampaignExecution> handleExecutionsForOneEnvironment(List<CampaignExecution> executionsFromOneEnvironment) {
-                campaignExecutionsByPartialExecution = executionsFromOneEnvironment.stream()
+                Map<Boolean, List<CampaignExecution>> campaignExecutionsByEmpty = executionsFromOneEnvironment.stream()
+                    .collect(groupingBy(cer -> cer.scenarioExecutionReports().isEmpty()));
+                emptyCampaignExecutions = ofNullable(campaignExecutionsByEmpty.get(true)).orElse(emptyList());
+
+                campaignExecutionsByPartialExecution = ofNullable(campaignExecutionsByEmpty.get(false)).orElse(emptyList()).stream()
                     .collect(groupingBy(cer -> cer.partialExecution));
-                return campaignExecutionsByPartialExecution.get(false);
+                return ofNullable(campaignExecutionsByPartialExecution.get(false)).orElse(emptyList());
             }
 
             /**
-             * Select all manual replays (for deletion) that are older than the oldest campaign execution kept.
+             * Select all manual replays (for deletion) that are older than the oldest campaign execution kept and empty executions.
              */
             @Override
             public Collection<Long> findExtraExecutionsIdsToDelete(List<CampaignExecution> timeSortedExecutionsForOneEnvironment) {
-                var oldestCampaignExecutionToKept = timeSortedExecutionsForOneEnvironment.get(maxCampaignExecutions - 1);
+                return Stream.concat(
+                    manualReplaysOlderThanOldestCampaignExecution(timeSortedExecutionsForOneEnvironment),
+                    emptyCampaignExecutions.stream().map(ce -> ce.executionId)
+                ).toList();
+            }
+
+            private Stream<Long> manualReplaysOlderThanOldestCampaignExecution(List<CampaignExecution> timeSortedExecutionsForOneEnvironment) {
+                LocalDateTime oldestCampaignExecutionToKeptStartDate;
+                if (!timeSortedExecutionsForOneEnvironment.isEmpty() && maxCampaignExecutions > 0) {
+                    oldestCampaignExecutionToKeptStartDate = timeSortedExecutionsForOneEnvironment.get(maxCampaignExecutions - 1).startDate;
```

Comment:
We never check that "timeSortedExecutionsForOneEnvironment.get(maxCampaignExecutions - 1)" exists